Clickhouse Trends Stickiness

mariusandra opened the pull request:
## Changes

Stickiness query rewritten for clickhouse.
Seems to work, though there's some code duplication with the trends part.

Also, I noticed the default time range is `"-14d"`, but the frontend considers the default to be `"-7d"`. Should probably be fixed to -7?

## Checklist

- [ ] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [ ] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

EDsCODE committed:
initial working

EDsCODE committed:
date param almost working

EDsCODE committed:
fix date range and labels

EDsCODE committed:
fixed monthly math

EDsCODE committed:
handle compare

EDsCODE committed:
Merge branch 'clickhouse-setup' into ch-trends-basic

EDsCODE committed:
change table

EDsCODE committed:
merge clickhouse-setup

EDsCODE committed:
using new event ingestion

EDsCODE committed:
Merge branch 'clickhouse-setup' into ch-trends-basic

EDsCODE committed:
Merge branch 'clickhouse-setup' into ch-trends-basic

EDsCODE committed:
Merge branch 'clickhouse-actions' into ch-trends-basic

EDsCODE committed:
direct query actions working

EDsCODE committed:
remove interface

EDsCODE committed:
fix date range

EDsCODE committed:
properties initial working

EDsCODE committed:
handle operator

EDsCODE committed:
handle operator

EDsCODE committed:
move timestamp parse

EDsCODE committed:
merge clickhouse-actions

EDsCODE committed:
move more to util

EDsCODE committed:
merge clickhouse-actions

EDsCODE committed:
inital breaking down working

mariusandra committed:
Fix actions with no steps crashing query

mariusandra committed:
Basic stickiness query

Twixes commented on ee/clickhouse/queries/clickhouse_stickiness.py:
```diff
@@ -1,10 +1,133 @@
+                return []
+
+            params = {**params, **action_params}
+            content_sql = STICKINESS_ACTIONS_SQL.format(
+                team_id=team.pk,
+                actions_query=action_query,
+                parsed_date_from=(parsed_date_from or ""),
+                parsed_date_to=(parsed_date_to or ""),
+                filters="AND id IN {filters}".format(filters=prop_filters) if filter.properties else "",
+            )
```
This seems susceptible to SQL injection attacks (particularly through filters). Do we have protections around that?

mariusandra replied:
It's definitely susceptible for this! I left it as is though because all other/existing clickhouse queries currently are equally vulnerable right now. This is definitely something to fix before any of this goes live.

EDsCODE replied:
yeah we will be doing a full parameterization before deploying. the syntax is just `%(variable_name)s`. i've done it in some spots but also just directly formatted in others out of convenience while i was experimenting

EDsCODE commented:
Could you add the tests to these? Sessions branch has the best example of how to do it and the command to run tests properly is 
```
python manage.py test <file path> --testrunner="ee.clickhouse.clickhouse_test_runner.ClickhouseTestRunner"
```

mariusandra committed:
Clickhouse Stickiness tests

mariusandra committed:
set CLICKHOUSE_SECURE=False by default if running in TEST or DEBUG

mariusandra committed:
change default test runner if PRIMARY_DB=clickhouse

mariusandra committed:
generate clickhouse uuid script

mariusandra committed:
convert person_id to UUID, make adding `person_id` optional, add distinct_ids already in the `create_person` function

mariusandra committed:
stickiness test [WIP, actions fail]

mariusandra committed:
Fix test_process_event_ee.py, remove all calls to Person.objects.*

mariusandra commented:
Hi! 

I did a bunch of stuff now... and I hope I didn't redo work that was already done in other branches :).

- Added the stickiness tests
  - Updated the stickiness test in general to also test that distinct_ids are merged properly into one person
  - The stickiness test with events works in clickhouse
  - The stickiness test with actions doesn't work yet (TODO)
- Updated the default test runner if PRIMARY_DB=clickhouse to not have to specify it on the command line
- Updated CLICKHOUSE_SECURE to be False if running in TEST/DEBUG
- Converted the person table ID field to UUID and updated references as needed
- Fixed the process_event EE test:
  - Scoped a bunch of clickhouse queries to team_id
  - Removed all `Person.objects.*` calls
  - Added `is_identified` to clickhouse

Still to do:
- Make the stickiness actions part work. I'm not sure what's the status with actions in clickhouse in general... so I thought to "save changes" before I start checking that...

mariusandra commented:
Also, before any of that I'll try to run all the tests and make sure nothing broke. Github is reporting some issues...

EDsCODE commented:
After looking at the changes I think it'd actually be best if you're able to split out the process event changes and get that merged to master independently sooner. These changes will be useful to @fuziontech for the event ingestion.

EDsCODE committed:
make sure route is registered

mariusandra commented:
@EDsCODE done here! https://github.com/PostHog/posthog/pull/1652

EDsCODE commented:
Note: rebase to `clickhouse-actions` and redirect the target branch to `clickhouse-actions`.

mariusandra commented:
Superseded by #1654
